File: python_playground/bubble.py
def sort(list_to_sort: list[int]) -> list[int]:
  sorted_count = 1
  while sorted_count != len(list_to_sort):
    swapped = False
    for i in range(0, len(list_to_sort) - sorted_count):
      if list_to_sort[i] > list_to_sort[i+1]:
        list_to_sort[i], list_to_sort[i+1] = list_to_sort[i+1], list_to_sort[i]
        swapped = True
    if not swapped:
      break
    sorted_count += 1
  
  return list_to_sort

File: algo_visualizer/src/algos/bubble.ts
import { BigO } from "./SortAlgorithm";
import type SortAlgorithm from "./SortAlgorithm";

export default class Bubble implements SortAlgorithm {
  readonly name = "Bubble Sort";
  readonly expectedBigO: BigO = BigO.N_SQUARED;
  readonly chartColor: string = "#ef4444";

  sort = (listToSort: number[]): number[] => {
    let sortedCount = 1;

    while (sortedCount !== listToSort.length) {
      let swapped = false;
      for (let i = 0; i < listToSort.length - sortedCount; i++) {
        if (listToSort[i] > listToSort[i + 1]) {
          [listToSort[i], listToSort[i + 1]] = [listToSort[i + 1], listToSort[i]];
          swapped = true;
        }
      }
      if (!swapped) {
        break;
      }
      sortedCount++;
    }

    return listToSort;
  };
}
